### PR TITLE
db/hints: Prevent dereferencing a null pointer

### PR DESCRIPTION
**Pull request by dawmd**
Before these changes, we dereferenced `app_state` in `manager::endpoint_downtime_not_bigger_than()` before checking that it's not a null pointer. We fix that.

Fixes scylladb/scylladb#21699

Backport: it's a serious issue, so we backport it to all maintained versions: 6.1 and 6.2.

### CODE REVIEW

**Review by piodul — APPROVED** [review]

File: db/hints/manager.cc [hunk 1]
```diff
@@ -492,9 +492,12 @@ static bool endpoint_downtime_not_bigger_than(const gms::gossiper& gossiper, con
 
     gossiper.for_each_endpoint_state_until(
             [&info] (const gms::inet_address& ip, const gms::endpoint_state& state) {
-        const auto app_state = state.get_application_state_ptr(gms::application_state::HOST_ID);
+        const auto* app_state = state.get_application_state_ptr(gms::application_state::HOST_ID);
+        if (!app_state) {
```
Comment (kbr-scylla) [inline db/hints/manager.cc:496]: let's print an ERROR if we detect a missing HOST ID. It should not happen, and separate issue is needed. Other than that LGTM

Comment (dawmd) [reply db/hints/manager.cc:496]: Good idea, thanks!

**Comment by dawmd** [comment]
Version 2:
* Print an error if the host ID application state is not found.

**Comment by avikivity** [comment]
@yaronkaikov docker test is failing everywhere, what happened?

**Comment by yaronkaikov** [comment]
> @yaronkaikov docker test is failing everywhere, what happened?

regression from https://github.com/scylladb/scylla-cluster-tests/pull/9331

@dimakr is on it

**Comment by dawmd** [comment]
Version 3:
* I decided to extend the log message by the endpoint state. I guess it might be helpful while debugging if the application state doesn't exist again.

**Comment by dawmd** [comment]
Version 4:
* I created a new issue and referenced it in the commit. The original one detected this problem, **but** it currently revolves around a more critical issue. I don't want to close it, so I go this way instead.

**Comment by piodul** [comment]
> #### Failed Tests (2/35888):
> * [test_kill_coordinator_during_op.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13399/testReport/junit/topology_experimental_raft/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op.debug.3)
> * [test_kill_coordinator_during_op.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13399/testReport/junit/topology_experimental_raft/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_debug_3_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op.debug.3)

Looks like https://github.com/scylladb/scylladb/issues/21114. Restarting CI

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Queued

**Pull request closed without merging**